syntax = "proto3";

package user;

service UserService {
    rpc GetUser(GetUserRequest) returns (UserResponse);
    rpc ValidateSubscription(ValidateSubscriptionRequest) returns (ValidationResponse);
    rpc CreateUser(CreateUserRequest) returns (UserResponse);
    rpc UpdateUser(UpdateUserRequest) returns (UserResponse);
    rpc GetUsersByType(GetUsersByTypeRequest) returns (UsersResponse);
}

message GetUserRequest {
    string user_id = 1;
}

message ValidateSubscriptionRequest {
    string user_id = 1;
}

message CreateUserRequest {
    string email = 1;
    string password = 2;
    string first_name = 3;
    string last_name = 4;
    string phone = 5;
    int32 type_id = 6;
    string subscription_end = 7;
}

message UpdateUserRequest {
    string user_id = 1;
    string email = 2;
    string first_name = 3;
    string last_name = 4;
    string phone = 5;
    string subscription_end = 6;
    string status = 7;
}

message GetUsersByTypeRequest {
    int32 type_id = 1;
}

message UserResponse {
    string user_id = 1;
    string email = 2;
    string first_name = 3;
    string last_name = 4;
    string phone = 5;
    int32 type_id = 6;
    string subscription_end = 7;
    string created_at = 8;
    string updated_at = 9;
    string status = 10;
}

message UsersResponse {
    repeated UserResponse users = 1;
}

message ValidationResponse {
    bool valid = 1;
    string message = 2;
}